avocado/plugins/resolvers.py

```diff
@@ -110,6 +110,25 @@ def resolve(reference):
                                find_avocado_tests)
 
 
+class RequirementsResolver:
+
+    name = 'requirements'
+    description = 'Requirements resolver for tests with requirements'
+
+    @staticmethod
+    def resolve(runnable):
+        requirements_runnables = []
+        for requirement in runnable.requirements:
+            requirement_copy = requirement.copy()
```

Comment:
The need to create a copy implies that you expect the original data to be modified.  Do you?  I can't think of how this data will come from the docstrings, and will then be modified.

Reply:
`type` is removed below. When removing the `type` from the original variable, it reflects in the runnable object and may become hard to create a consistent log in the end.

Reply:
> `type` is removed below. When removing the `type` from the original variable, it reflects in the runnable object and may become hard to create a consistent log in the end.

The fact that you're stripping parts of the original object is not be a valid reason for a copy.  Like I said before, if the *original* data structure is not expected to be modified, it shouldn't be copied.  You can do the same thing without a copy:

```python
    @staticmethod
    def resolve(runnable):
        requirements_runnables = []
        for requirement in runnable.requirements:
            kind = 'requirement-%s' % requirement['type']
            package = requirement['name']
            requirement_runnable = Runnable(kind, None, package)
            # If, it's expected that a list can be given for package *names*, then
            # special treatment should be given here, such as:
            # >>> packages = ast.literal_eval(requirement['names'])
            # >>> requirement_runnable = Runnable(kind, None, *packages)

            # for now, ignore unsupported requirements types
            if requirement_runnable.pick_runner_command() is not None:
                requirements_runnables.append(requirement_runnable)
        return requirements_runnables
```

Reply:
Oh, I see now that I may not have explained this block with a full context, and you have just the package requirement in mind. The code I proposed already supports other future implementations of different requirement types. For example, an asset defined like:

```
{"type": "file", "name": "cabort", "uri": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}
```

The code removes the `type` and passes everything else to the runner. The runner knows what arguments are expected and handles them accordingly.

With the package requirement, if there are arguments that do not mean anything to the `type`, they are just ignored, like, for example:

```
{"type": "package", "name": "vim-common", "foo": "bar"}
```

The package runner ignores the `foo` argument because it does not have an implementation to handle it.

As the number of arguments may vary from `type` to `type`, this implementation does not need to worry about it and let the runner handle the arguments.

About the `ast` solution in the comment, I don't see an advantage in using a module to create a list of arguments if we already have the `key:value` dict and just need to remove the `type` key from it.

I hope the examples I gave make things more clear now.

Reply:
> Oh, I see now that I may not have explained this block with a full context, and you have just the package requirement in mind. The code I proposed already supports other future implementations of different requirement types. For example, an asset defined like:
> 
> ```
> {"type": "file", "name": "cabort", "uri": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}
> ```
> 
> The code removes the `type` and passes everything else to the runner. The runner knows what arguments are expected and handles them accordingly.
> 
> With the package requirement, if there are arguments that do not mean anything to the `type`, they are just ignored, like, for example:
> 
> ```
> {"type": "package", "name": "vim-common", "foo": "bar"}
> ```
> 
> The package runner ignores the `foo` argument because it does not have an implementation to handle it.
> 
> As the number of arguments may vary from `type` to `type`, this implementation does not need to worry about it and let the runner handle the arguments.
> 

I see.  But for this mapping to be always automatic and transparent like you suggest, all arguments would then need to be keyword based.  I wonder if that makes sense for the `requirement-package` case.  Like I said before, it looks to me like:

```
    * args: package names to act upon
```

Would make this automatic mapping impossible.

> About the `ast` solution in the comment, I don't see an advantage in using a module to create a list of arguments if we already have the `key:value` dict and just need to remove the `type` key from it.
> 

Sure, that was not meant to be taken literally.  My point is that, if we determine that `name` can be a list, then it should always be a list -- either transparently from JSON or say by parsing a string as a list.

> I hope the examples I gave make things more clear now.